File: amazon/42_trapping_rain_water.py

#two pointers approach.
class Solution:
    def trap(self, height: List[int]) -> int:

        if len(height) <= 2:
            return 0

        ans = 0
        #using two pointers i and j on indices 1 and n - 1
        i = 1
        j = len(height)-1
        #intialising leftmax to the leftmost bar and rightmax to the rightmost bar
        left_max = height[0]
        right_max = height[-1]

        while i <= j:
            #check lmax and rmax for current i, j positions
            if height[i] > left_max:
                left_max = height[i]
            if height[j] > right_max:
                right_max = height[j]
            #fll water up to lmax level for index i and move i to the right
            if left_max <= right_max:
                ans += left_max - height[i]
                i += 1
#fill water up to rightmax level for index j and move j to the left
            else:
                ans += right_max - height[j]
                j -= 1

        return ans

            

File: amazon/42_trapping_rain_water.cpp
class Solution{
public:
    int trap(vector<int>& height){
        int water = 0;
        int capacity_tank = height.size();
        int left[capacity_tank];
        int right[capacity_tank];
        left[0] = height[0];
        right[capacity_tank - 1] = height[capacity_tank-1];
        for(int i = 1; i < capacity_tank; i++){
            left[i] = max(height[i], right[i-1]);
        }
        for(int i = capacity_tank -2; i > 0; i--){
            right[i] = max(height[i], right[i+1]);
        }
        for(int i = 1; i < capacity_tank; i++){
            water += min(left[i], right[i]) - height[i];
        }
        return water;
    }
}